mat64/eigen.go

```diff
@@ -798,22 +800,23 @@ func hqr2(d, e []float64, hess, v *Dense, epsilon float64) {
 	}
 }
 
-// D returns the block diagonal eigenvalue matrix from the real and imaginary
-// components d and e.
-func (f EigenFactors) D() *Dense {
-	d, e := f.d, f.e
-	var n int
-	if n = len(d); n != len(e) {
-		panic(ErrSquare)
-	}
-	dm := NewDense(n, n, nil)
-	for i := 0; i < n; i++ {
-		dm.set(i, i, d[i])
-		if e[i] > 0 {
-			dm.set(i, i+1, e[i])
-		} else if e[i] < 0 {
-			dm.set(i, i-1, e[i])
+type EigenValues []complex128
+
+// D returns the eigenvalue matrix from the real and imaginary components
+// of the eigenvalues.
+//
+// If the decomposed matrix is not symmetric, then D returns a block diagonal
+// matrix with the real eigenvalues in 1-by-1 blocks and any complex eigenvalues,
+// lambda + i*mu, in 2-by-2 blocks, [lambda, mu; -mu, lambda].
+func (v EigenValues) D() *Dense {
```

Comment:
Interestingly, the eigenvectors I get from Matlab don't agree with the eigenvalues you get from matlab
````
> a = [1,6,7; 9,1,10; 15,6,1]

a =

     1     6     7
     9     1    10
    15     6     1

> [v,d] = eig(a)

v =

   0.4730             0.1471 - 0.1444i   0.1471 + 0.1444i
   0.6161             0.6380 + 0.2330i   0.6380 - 0.2330i
   0.6298            -0.7044            -0.7044          


d =

  18.1344                  0                  0          
        0            -7.5672 + 1.0909i        0          
        0                  0            -7.5672 - 1.0909i
````
In the end, we can only agree up to an extent.

The bigger issue is that the mat64 eigenvectors are not orthonormal, and not only that, appear to not even be eigenvalues

````
> a = [1,6,7; 9,1,10; 15,6,1]

a =

     1     6     7
     9     1    10
    15     6     1
> [v,d ] = eig(a)

v =

   0.4730             0.1471 - 0.1444i   0.1471 + 0.1444i
   0.6161             0.6380 + 0.2330i   0.6380 - 0.2330i
   0.6298            -0.7044            -0.7044          


d =

  18.1344                  0                  0          
        0            -7.5672 + 1.0909i        0          
        0                  0            -7.5672 - 1.0909i

> v * d * inv(v)

ans =

   1.0000 + 0.0000i   6.0000 - 0.0000i   7.0000 - 0.0000i
   9.0000 + 0.0000i   1.0000 - 0.0000i  10.0000 - 0.0000i
  15.0000 + 0.0000i   6.0000 - 0.0000i   1.0000 - 0.0000i

> v2 = [0.47304093468892405  -0.6629102254940207   0.6029367060177389; 0.6160598119673665  -2.7326877058673555   -1.117348566292792; 0.6298432997082382    3.059661412350093  0.11648819407649919]

v2 =

    0.4730   -0.6629    0.6029
    0.6161   -2.7327   -1.1173
    0.6298    3.0597    0.1165

> v2 * d * inv(v2)

ans =

   1.5049 - 0.4359i   5.6240 + 0.3517i   6.9886 - 0.0167i
  11.8149 + 1.6144i  -0.2428 - 0.3142i   9.1015 - 0.9052i
  12.0793 - 0.7333i   7.4882 - 0.2038i   1.7379 + 0.7501i
````